File: packages/python_host/src/one_sdk/sf_host.py
from typing import TYPE_CHECKING, Tuple
if TYPE_CHECKING:
	from one_sdk.app import WasiApp

import sys
import json
import functools

from wasmtime import Linker, FuncType, ValType

from one_sdk.handle_map import HandleMap
from one_sdk.error import WasiError, WasiErrno

# TODO: TypeAlias - needs 3.10
Ptr = int
Size = int
AbiResult = int

def _strace_inner(fn, name, *args):
	result = fn(*args)
	print(f"host: [strace] {name}{args} = {result}", file = sys.stderr)
	return result
def strace(fn, name):
	"""Use on a function to wrap with a debug print when called."""
	return functools.partial(_strace_inner, fn, name)

def _join_abi_result(lower: int, upper: int) -> AbiResult:
	lower = lower & 0x7FFFFFFF
	upper = (upper & 0x1) << 31

	return lower | upper
def _split_abi_result(value: AbiResult) -> Tuple[int, int]:
	lower = value & 0x7FFFFFFF
	upper = (value >> 31) & 0x1

	return (lower, upper)
def _abi_ok(value: int) -> AbiResult:
	return _join_abi_result(value, 0)
def _abi_err(value: int) -> AbiResult:
	return _join_abi_result(int(value), 1)

def link(linker: Linker, app: "WasiApp"):
	message_store: HandleMap[bytes] = HandleMap()

	def __export_message_exchange(msg_ptr: Ptr, msg_len: Size, out_ptr: Ptr, out_len: Size, ret_handle: Ptr) -> Size:
		memory = app.memory
		message = json.loads(
			memory.read_bytes(msg_ptr, msg_len).decode("utf-8")
		)
		response_bytes = json.dumps(app.handle_message(message)).encode("utf-8")

		message_handle = 0
		if len(response_bytes) > out_len:
			message_handle = message_store.insert(response_bytes)
		else:
			memory.write_bytes(out_ptr, out_len, response_bytes)
		
		memory.write_i32(ret_handle, message_handle)
		return _abi_ok(len(response_bytes))

	def __export_message_exchange_retrieve(handle: int, out_ptr: Ptr, out_len: Size) -> AbiResult:
		response_bytes = message_store.remove(handle)
		if response_bytes is None:
			return _abi_err(WasiErrno.EBADF)
		if out_len < len(response_bytes):
			return _abi_err(WasiErrno.ERANGE)
		
		count = app.memory.write_bytes(out_ptr, out_len, response_bytes)
		return _abi_ok(count)

	def __export_stream_read(handle: int, out_ptr: Ptr, out_len: Size) -> AbiResult:
		try:
			read_count = app.memory.write_bytes(out_ptr, out_len, app.stream_read(handle, out_len))
		except WasiError as e:
			return _abi_err(e.errno)
		
		return _abi_ok(read_count)
	
	def __export_stream_write(handle: int, in_ptr: Ptr, in_len: Size) -> AbiResult:
		try:
			write_count = app.stream_write(handle, app.memory.read_bytes(in_ptr, in_len))
		except WasiError as e:
			return _abi_err(e.errno)
		
		return _abi_ok(write_count)

	def __export_stream_close(handle: int) -> AbiResult:
		try:
			app.stream_close(handle)
		except WasiError as e:
			return _abi_err(e.errno)
		
		return _abi_ok(WasiErrno.SUCCESS)
	
	linker.define_func(
		"sf_host_unstable", "message_exchange",
		FuncType([ValType.i32(), ValType.i32(), ValType.i32(), ValType.i32(), ValType.i32()], [ValType.i32()]),
		__export_message_exchange
	)
	linker.define_func(
		"sf_host_unstable", "message_exchange_retrieve",
		FuncType([ValType.i32(), ValType.i32(), ValType.i32()], [ValType.i32()]),
		__export_message_exchange_retrieve
	)
	linker.define_func(
		"sf_host_unstable", "stream_read",
		FuncType([ValType.i32(), ValType.i32(), ValType.i32()], [ValType.i32()]),
		__export_stream_read
	)
	linker.define_func(
		"sf_host_unstable", "stream_write",
		FuncType([ValType.i32(), ValType.i32(), ValType.i32()], [ValType.i32()]),
		__export_stream_write
	)
	linker.define_func(
		"sf_host_unstable", "stream_close",
		FuncType([ValType.i32()], [ValType.i32()]),
		__export_stream_close
	)

File: packages/javascript_common/src/lib/sf_host.ts
import type { TextCoder, AppContext, AppContextSync } from './interfaces.js';
import { Asyncify, AsyncifyState } from './asyncify.js'
import { HandleMap } from './handle_map.js'
import { WasiErrno } from './error.js';

type AbiResult = number;
// @ts-ignore ignore the unused parameter
type Ptr<T> = number;
type Size = number;

type Fn<A extends unknown[], R> = (...args: A) => R;
function strace<A extends unknown[], R>(name: string, fn: Fn<A, R>, asyncify: Asyncify): Fn<A, R> {
  return (...args: A): R => {
    const result: any = fn(...args);

    // TODO: control this from variable
    if (asyncify.getState() === AsyncifyState.Normal) {
      console.debug(`host: [strace] ${name}(${args}) = ${result}`);
    } else {
      console.debug(`host: [strace] ${name}(${args}) = <async suspended (${result})>`);
    }

    return result;
  }
}
// @ts-ignore ignore unused function
function strace_module(moduleName: string, module: WebAssembly.ModuleImports, asyncify: Asyncify): WebAssembly.ModuleImports {
  for (const fnName of Object.keys(module)) {
    const fn: any = module[fnName];
    const name = `${moduleName}::${fnName}`;
    module[fnName] = strace(name, fn, asyncify);
  }

  return module;
}

function join_abi_result(lower: number, upper: number): number {
  const l = lower & 0x7FFFFFFF;
  const u = (upper & 0x1) << 31;

  return l | u;
}
// @ts-ignore ignore unused funtion
function split_abi_result(value: number): [lower: number, upper: number] {
  const lower = value & 0x7FFFFFFF;
  const upper = (value >> 31) & 0x1;

  return [lower, upper];
}
function abi_ok(value: number): number {
  return join_abi_result(value, 0);
}
function abi_err(value: number): number {
  return join_abi_result(value, 1);
}

function writeBytes(source: ArrayBuffer, dest: Uint8Array) {
  const src = new Uint8Array(source);
  for (let i = 0; i < src.byteLength; i += 1) {
    dest[i] = src[i];
  }
}

export function link(app: AppContext, textCoder: TextCoder, asyncify: Asyncify): WebAssembly.Imports {
  const messageStore = new HandleMap<ArrayBuffer>();

  async function __export_message_exchange(msg_ptr: Ptr<8>, msg_len: Size, out_ptr: Ptr<8>, out_len: Size, ret_handle: Ptr<32>): Promise<Size> {
    const msg = JSON.parse(textCoder.decodeUtf8(
      app.memoryBytes.subarray(msg_ptr, msg_ptr + msg_len)
    ));
    const response = await app.handleMessage(msg);

    let messageHandle = 0;
    const responseBytes = textCoder.encodeUtf8(JSON.stringify(response));
    if (responseBytes.byteLength > out_len) {
      messageHandle = messageStore.insert(responseBytes);
    } else {
      const out = app.memoryBytes.subarray(out_ptr, out_ptr + out_len);
      writeBytes(responseBytes, out);
    }

    app.memoryView.setInt32(ret_handle, messageHandle, true);
    return responseBytes.byteLength;
  }
  function __export_message_exchange_retrieve(handle: number, out_ptr: Ptr<8>, out_len: Size): AbiResult {
    const responseBytes = messageStore.remove(handle);
    if (responseBytes === undefined) {
      return abi_err(WasiErrno.EBADF);
    }
    if (out_len < responseBytes.byteLength) {
      return abi_err(WasiErrno.ERANGE);
    }

    const out = app.memoryBytes.subarray(out_ptr, out_ptr + out_len);
    writeBytes(responseBytes, out);

    return abi_ok(responseBytes.byteLength);
  }

  async function __export_stream_read(handle: number, out_ptr: Ptr<8>, out_len: Size): Promise<AbiResult> {
    const out = app.memoryBytes.subarray(out_ptr, out_ptr + out_len);

    return app.readStream(handle, out).then(c => abi_ok(c), e => abi_err(e.errno));
  }
  async function __export_stream_write(handle: number, in_ptr: Ptr<8>, in_len: Size): Promise<AbiResult> {
    const data = app.memoryBytes.subarray(in_ptr, in_ptr + in_len);

    return app.writeStream(handle, data).then(c => abi_ok(c), e => abi_err(e.errno));
  }
  async function __export_stream_close(handle: number): Promise<AbiResult> {
    return await app.closeStream(handle).then(_ => abi_ok(0), e => abi_err(e.errno))
  }

  return {
    sf_host_unstable: {
      message_exchange: asyncify.wrapImport(__export_message_exchange, 0),
      message_exchange_retrieve: __export_message_exchange_retrieve,
      stream_read: asyncify.wrapImport(__export_stream_read, 0),
      stream_write: asyncify.wrapImport(__export_stream_write, 0),
      stream_close: asyncify.wrapImport(__export_stream_close, 0)
    }
  }
}

export function linkSync(app: AppContextSync, textCoder: TextCoder): WebAssembly.Imports {
  const messageStore = new HandleMap<ArrayBuffer>();

  function __export_message_exchange(msg_ptr: Ptr<8>, msg_len: Size, out_ptr: Ptr<8>, out_len: Size, ret_handle: Ptr<32>): Size {
    const msg = JSON.parse(textCoder.decodeUtf8(
      app.memoryBytes.subarray(msg_ptr, msg_ptr + msg_len)
    ));
    const response = app.handleMessage(msg);

    let messageHandle = 0;
    const responseBytes = textCoder.encodeUtf8(JSON.stringify(response));
    if (responseBytes.byteLength > out_len) {
      messageHandle = messageStore.insert(responseBytes);
    } else {
      const out = app.memoryBytes.subarray(out_ptr, out_ptr + out_len);
      writeBytes(responseBytes, out);
    }

    app.memoryView.setInt32(ret_handle, messageHandle, true);
    return responseBytes.byteLength;
  }
  function __export_message_exchange_retrieve(handle: number, out_ptr: Ptr<8>, out_len: Size): AbiResult {
    const responseBytes = messageStore.remove(handle);
    if (responseBytes === undefined) {
      return abi_err(WasiErrno.EBADF);
    }
    if (out_len < responseBytes.byteLength) {
      return abi_err(WasiErrno.ERANGE);
    }

    const out = app.memoryBytes.subarray(out_ptr, out_ptr + out_len);
    writeBytes(responseBytes, out);

    return abi_ok(responseBytes.byteLength);
  }

  function __export_stream_read(handle: number, out_ptr: Ptr<8>, out_len: Size): AbiResult {
    const out = app.memoryBytes.subarray(out_ptr, out_ptr + out_len);

    try {
      const c = app.readStream(handle, out)
      return abi_ok(c)
    } catch (e: any) {
      return abi_err(e.errno)
    }
  }
  function __export_stream_write(handle: number, in_ptr: Ptr<8>, in_len: Size): AbiResult {
    const data = app.memoryBytes.subarray(in_ptr, in_ptr + in_len);

    try {
      const c = app.writeStream(handle, data)
      return abi_ok(c)
    } catch (e: any) {
      return abi_err(e.errno)
    }
  }
  function __export_stream_close(handle: number): AbiResult {
    try {
      app.closeStream(handle)
      return abi_ok(0)
    } catch (e: any) {
      return abi_err(e.errno)
    }
  }

  return {
    sf_host_unstable: {
      message_exchange: __export_message_exchange,
      message_exchange_retrieve: __export_message_exchange_retrieve,
      stream_read: __export_stream_read,
      stream_write: __export_stream_write,
      stream_close: __export_stream_close,
    }
  }
}
